src/sender.rs

```diff
@@ -390,16 +388,12 @@ impl<'a, 'b, T> Wait<'a, 'b, T> {
 
 #[cfg(all(feature="async", not(feature="messy")))]
 impl<'a, 'b, T> Drop for Wait<'a, 'b, T> {
-    // clean out the waker. essentially this makes our benchmarks worse but should improve real
-    // world performance, since we assume that an async executor can't match our performance.
     fn drop(&mut self) {
-        if WAITING != (self.flags | self.sender.flags) & (S_CLOSE | WAITING) { return; }
-        // We'll need a lock as usual
+        if WAITING != (self.flags | self.sender.flags) & (S_CLOSE | WAITING) { return; } // Nothing to do.
         let flags = self.sender.lock();
-        // Our cleanup is to remove the waker. Also release the lock.
         let shared = unsafe { &mut *self.sender.hatch.inner.get() };
```

Comment:
Safety remarks?

Reply:
It seemed a little excessive given the lock is taken in the previous line.

Reply:
ah ok, so they are directly related, e.g. `sender.lock` locks the hatch?

Reply:
yes. the  spinlock (taken via `sender.lock` / `receiver.lock`) provides mutual exclusion for the value and wakers (in the async case).

we assume that the other side has always made a change (acquire) and we release our changes when we unlock.

Reply:
I know we used to have individual locks, but it didn't seem to be winning us anything while it was increasing complexity. now we just have the one lock for all the mutable hatch data.

the `sender.lock` / `receiver.lock` return either when they have the lock or they detect the other side has closed (by reading the returned flags). in the latter case, they will update the sender or receiver's local copy of the flags, allowing us to elide further atomics.

Reply:
Is this faster than using an implementation based on `parking-lot` mutexes?

Reply:
a mutex is the wrong primitive, it implies that on failure (usually after a few spins) it will use a CondVar to get the OS to sleep the current thread. That is the absolute last thing we want in an async environment because it is just asking for the executor to be held up.

so we do the next best thing, a spinlock.

Reply:
actually, you accidentally stumbled upon something quite interesting about the current design of async rust - there is no better way to handle this than a spin loop.

imagine the sender and receiver are on different threads. the sender takes the lock and then the OS deschedules it, then the receiver tries to take the lock. that spinlock is not going to succeed until the sender thread gets rescheduled.

in a non-async environment, we would actually want the OS to be involved, but in an async environment, that is not acceptable.

i have mooted previously that it would be nice if you could ask to reschedule the current async task. There is a hack for doing that (wake yourself, return pending), but it only does a reasonable job if you have fair scheduling (e.g. tokio's scheduler uses a 'next' slot which would just cause the same future to be executed again in all likelihood. it's a common optimisation and the workaround (also scheduling a dummy) would penalise users of executors that *do* provide fair queuing if we didn't lock it behind (yet another) feature flag. also it's just horrible and i'm hopeful that eventually we can get better support from the rust stdlib.

Reply:
It is definitively possible to implement a mutex that is compatible with `async`, although it definitively takes a bit more work... Have you already looked at [`async-lock`](https://docs.rs/async-lock)?

Reply:
no, it's not.

because ultimately, you have to do whatever it is you're doing or set a waker to be woken up again. but the lock construct sets a waker! under the hood, even async-waker is a spinlock.

Reply:
Additionally, we'd no longer be able to squeeze all our flags down to a single atomic if we used an existing implementation